components/TARGET_PSA/services/crypto/COMPONENT_SPE/psa_crypto_partition.c

```diff
@@ -1607,39 +1638,36 @@ void psa_crypto_generator_operations(void)
                     }
 
                     salt = mbedtls_calloc(1, msg.in_size[1]);
-                    if (salt == NULL) {
+                    label = mbedtls_calloc(1, msg.in_size[2]);
+                    if (salt == NULL || label == NULL) {
                         status = PSA_ERROR_INSUFFICIENT_MEMORY;
-                        break;
-                    }
+                    } else {
+                        bytes_read = psa_read(msg.handle, 1, salt, msg.in_size[1]);
+                        if (bytes_read != msg.in_size[1]) {
+                            SPM_PANIC("SPM read length mismatch");
+                        }
 
-                    bytes_read = psa_read(msg.handle, 1, salt,
-                                          msg.in_size[1]);
-                    if (bytes_read != msg.in_size[1]) {
-                        SPM_PANIC("SPM read length mismatch");
-                    }
+                        bytes_read = psa_read(msg.handle, 2, label, msg.in_size[2]);
+                        if (bytes_read != msg.in_size[2]) {
+                            SPM_PANIC("SPM read length mismatch");
+                        }
 
-                    uint8_t *label = mbedtls_calloc(1, msg.in_size[2]);
-                    if (label == NULL) {
-                        status = PSA_ERROR_INSUFFICIENT_MEMORY;
-                        mbedtls_free(salt);
-                        break;
-                    }
+                        status = psa_key_derivation(msg.rhandle, psa_crypto_ipc.handle,
+                                                    psa_crypto_ipc.alg,
+                                                    salt,
+                                                    msg.in_size[1],//salt length
+                                                    label,
+                                                    msg.in_size[2],//label length
+                                                    psa_crypto_ipc.capacity);
 
-                    bytes_read = psa_read(msg.handle, 2, label,
-                                          msg.in_size[2]);
-                    if (bytes_read != msg.in_size[2]) {
-                        SPM_PANIC("SPM read length mismatch");
                     }
 
-                    status = psa_key_derivation(msg.rhandle, psa_crypto_ipc.handle,
-                                                psa_crypto_ipc.alg,
-                                                salt,
-                                                msg.in_size[1],//salt length
-                                                label,
-                                                msg.in_size[2],//label length
-                                                psa_crypto_ipc.capacity);
-                    mbedtls_free(label);
                     mbedtls_free(salt);
+                    mbedtls_free(label);
+                    if (status != PSA_SUCCESS) {
```

Comment:
this block returns dozens of times.
please wrap in in static cleanup named function.
i would also suggest considering instead of repeating it multiple times through various cases. considering some bailout flag and logic and doing cleanup related to this flag once before return or so.
code becomes now a bit spaghetti like and is being a bit blown in size

Reply:
I'll provide a static cleanup function. I don't like the error flag approach in this case because it will add even more complexity to these nested `switch` cases. We'll refactor the partition also once we get a chance.